language/stdlib/modules/approved_payment.move

```diff
@@ -0,0 +1,97 @@
+// Module that allows a payee to approve payments with a cryptographic signature. The basic flow is:
+// (1) Payer sends `metadata` to the payee
+// (2) Payee signs `metadata` and sends 64 byte signature back to the payer
+// (3) Payer sends an approved payment to the payee by sending a transaction invoking `deposit`
+//     with payment metadata + signature. The transaction will abort if the signature check fails.
+// Note: approved payments are an accounting convenience/a courtesy mechansim for the payee, *not*
+// a hurdle that must be cleared for all payments to the payee. In addition, approved payments do
+// not have replay protection.
+address 0x0:
+module ApprovedPayment {
+    use 0x0::Libra;
+    use 0x0::LibraAccount;
+    use 0x0::Signature;
+    use 0x0::Transaction;
+    use 0x0::Vector;
+
+    // A resource to be published under the payee's account
+    resource struct T {
+        // 32 byte single Ed25519 public key whose counterpart must be used to sign the payment
+        // metadata. Note that this is different (and simpler) than the `authentication_key` used in
+        // LibraAccount::T, which is a hash of a public key + signature scheme identifier.
+        public_key: vector<u8>,
+        // TODO: events?
+    }
+
+    // Deposit `coin` in `payee`'s account if the `signature` on the payment metadata matches the
+    // public key stored in the `approved_payment` resource
+    public fun deposit<Token>(
+        approved_payment: &T,
+        payee: address,
```

Comment:
Super nit (feel free to ignore): Other deposit functions in `LibraAccount` usually take the payee address as the first argument and the coin as second, followed by everything else. Was wondering if we wanted to also keep that convention here?

Reply:
Good question--I was actually basing this on some code in `Libra` that implements the "support doing X on both a top-level resource and a resource stored in a field" pattern. That is:

```
do_X(t: &mut T, args) { // actually do X }
do_X_from_sender(args) { do_X(borrow_global_mut<T>(Transaction::sender(), args) }
```

I think for this pattern, it makes sense to put `&T` first and the args second to make it clear how the signatures are different.